### ISSUE TITLE
Support for the {n,m} form for property paths

### ISSUE THREAD

**Issue by jerdebsp** [w3c/data-shapes#195]
In SPARQL you can have a complex property path with the {n,m} form e.g `?otherClass (sh:property/sh:class){2,3} ?nodeShape .` or the {n} form `?otherClass (sh:property/sh:class){2} ?nodeShape .`.

Is it possible to add this to the SHACL property path for property shapes?

Thanks,
Jer

**Comment by HolgerKnublauch** [comment]
Yes I think this was requested by others too, and would potentially make sense for SHACL 1.2 for symmetry reasons with SPARQL alone.

**Comment by TallTed** [comment]
Note that the `{n,m}` property path notation is *not* in SPARQL 1.0 nor SPARQL 1.1. It *was* in a *draft* of SPARQL 1.1, and has been raised as [potentially a part of a future SPARQL version](https://github.com/w3c/sparql-dev/issues/101). It has also been raised specifically [regarding the SPARQL 1.2 specification now in development](https://github.com/w3c/sparql-query/issues/147).

**Comment by VladimirAlexiev** [comment]
@jerdebsp Can you list some realistic use cases that use such a feature?

**Comment by ajnelson-nist** [comment]
Would these boundaries be implemented with extra predicates on the node that is the subject of `sh:oneOrMorePath` or `sh:zeroOrMorePath`?  Or would a third predicate be used instead, something like `sh:pathInRange`?

**Comment by jerdebsp** [comment]
@VladimirAlexiev sorry I missed this notification! We used it to get a particular sub-graph starting from a node and we only wanted to traverse up to a particular number of hops. A simple example could be a family tree - I want to get only up to 2nd generation of people from this particular node.

@ajnelson-nist I think it should be `sh:pathInRange` because `sh:zeroOrMorePath` would be equivalent to {0,m} and the `sh:oneOrMorePath` {1,m}, we might want to do {4,7} for example. Most probably the hardest part would be how engines will implement this efficiently

Other possibility could be that instead of having functionality and predicates, we can have some abstract property that would need to have it's implementation plugged in the SHACL engine - similar to Jena's Abstract Functions. @HolgerKnublauch what do you think? That would ensure that the specs will only have what's commonly used, but on the other hand we can have additional predicates and functionality that can be shared (DASH comes into my mind - DASH + engine specific implementation?)

**Comment by HolgerKnublauch** [comment]
See 3. on https://github.com/w3c/data-shapes/issues/215 I suggest to generalize sh:path to allow any Node Expression. Then see if the existing node expression types are sufficient to express what you need here, otherwise add a new expression type.

**Comment by ajnelson-nist** [comment]
> [@ajnelson-nist](https://github.com/ajnelson-nist) I think it should be `sh:pathInRange` because `sh:zeroOrMorePath` would be equivalent to {0,m} and the `sh:oneOrMorePath` {1,m}, we might want to do {4,7} for example. Most probably the hardest part would be how engines will implement this efficiently

I'm willing to think out loud about the alternative to @HolgerKnublauch 's generalization-of-`sh:path` expression a little bit more.  Hopefully seeing a potential implementation of the alternative written down will clarify whether generalized `sh:path` is the group preference.

@jerdebsp : I see a single predicate `sh:pathInRange` leading to introducing a step of grammar specification and string parsing for just this one property.  E.g., which of these would be valid?

* ` [] sh:pathInRange '{4,7}' . `
* ` [] sh:pathInRange '4,7' . `
* ` [] sh:pathInRange '{-7,-4}' . `  (I could see people comfortable with Python's negative slice indexing advocating for this to suggest inverse paths.)
* ` [] sh:pathInRange '{-1,1}' . `
* ` [] sh:pathInRange '{4,aaaaaaaaaaaa' . `

I would prefer a pair of properties, with at least one being required to be provided.  I'm not attached to these suggested property names.

```turtle
ex:SomePropertyShape
  a sh:PropertyShape ;
  sh:path (
      [
        sh:zeroOrMorePath ex:a ;
        sh:atLeastPath 4 ;
        sh:atMostPath 7 ;
      ]
      [
        sh:oneOrMorePath ex:b ;
        sh:atLeastPath 0 ;
        sh:atMostPath 3 ;
      ]
    ) ;
.
```

We hadn't explicitly dived into it, but if this design were adopted, there's an argument for SHACL-SHACL being updated to flag that path component about `ex:b` with a `sh:Warning` about the lower bound of 0 being weird next to `sh:oneOrMorePath`.

Meanwhile, kind of a side note, I just thought of it while sketching the string examples - would the range of these properties need to be bounded to non-negative integers?  I followed @TallTed 's link a bit, but didn't see number bounds specifically, unless they were bundled into the discussion on counting alluded to in [SEP-0003](https://github.com/w3c/sparql-dev/blob/6637c7cbb670c0a19d2391b966df1168de127cb7/SEP/SEP-0003/sep-0003.md) (current state linked).

**Comment by jerdebsp** [comment]
Yes I agree. Holger's suggestion is neat and less ambiguous

**Comment by ajnelson-nist** [comment]
@jerdebsp , I'm not as read up on the node expressions Holger had linked.  Are you able to have a try at using that solution to model that weird property shape I'd done?  I.e., adapting this SPARQL (purposefully avoiding `{m,n}`, since apparently that isn't yet a standard syntax)?

```sparql
SELECT ?focus
WHERE {
#       1    2    3    4    5?    6?    7?    1?    2?    3?
  $this ex:a/ex:a/ex:a/ex:a/ex:a?/ex:a?/ex:a?/ex:b?/ex:b?/ex:b? ?focus .
}
```

**Comment by HolgerKnublauch** [comment]
I suggest closing this ticket. The use cases can probably be handled by sh:targetNode node expressions, and I see little appetite to change the sh:path syntax for a feature that isn't even in SPARQL 1.2

(We need to do house keeping on open tickets as the WG is bogged down right now)

**Comment by afs** [comment]
I agree we need to have two streams of open issues which are work items and open issues are a record of feature requests.

Would having a new repo, primarily for its issues list, work? It is easy to transfer issues between repos.

e.g. https://github.com/w3c/shacl-cg presuming a CG post-WG.

While a label might help (there is already "Postponed" but maybe "Feature Request" or "After WG") it would mean closing issues, or remembering to filter correctly if left open, neither of which are friendly.

(agenda item added)

**Comment by HolgerKnublauch** [comment]
In the meeting today we suggested to establish a procedure that any Core or SPARQL tickets that are not assigned to anyone (whether editor or not) by the end of June will be marked as postponed and the core or sparql labels will be removed. This would take them off the list without being forgotten, and if people have time they can still work on them. But it would allow us to set a clear end date so that the WG can officially move to the next phase.

**Comment by TallTed** [comment]
Note that multiple labels may be applied to each issue, so we can say both "Feature Request" *and* "After WG", so we can indicate both the kind of each issue *and* our planned timing for handling that issue.

**Comment by HolgerKnublauch** [comment]
Marking this as postponed as we are beyond the planned work cycle for SHACL Core. We can revisit this if there is renewed time and interest.
